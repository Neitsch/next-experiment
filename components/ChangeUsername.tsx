import React from "react";
import TextField from "@material-ui/core/TextField";
import gql from "graphql-tag";
import { graphql, ChildProps } from "react-apollo";

import {
  ChangeUsernameVariables,
  ChangeUsername,
} from "../query-types/ChangeUsername";

interface IProps {}

type Props = ChildProps<IProps, ChangeUsername, ChangeUsernameVariables>;

type State = {
  username: string;
};

class ChangeUsernameComponent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      username: "",
    };
  }
  public render() {
    console.log(this.props);
    if (this.props.data && this.props.data.error) {
      console.log(this.props.data.error);
    }
    return (
      <TextField
        onChange={this.usernameChange}
        onKeyPress={this.usernameSubmit}
      />
    );
  }
  private usernameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const username = e.currentTarget.value;
    this.setState(() => ({
      username,
    }));
  };
  private usernameSubmit = (e: React.KeyboardEvent<HTMLDivElement>) => {
    if (e.charCode == 13 && this.props.mutate) {
      this.props.mutate({ variables: { username: this.state.username } });
    }
  };
}

const changeUsername = gql`
  mutation ChangeUsername($username: String!) {
    changeUsername(username: $username) {
      username
      id
    }
  }
`;

export default graphql<Props, ChangeUsername, ChangeUsernameVariables>(
  changeUsername,
)(ChangeUsernameComponent);
