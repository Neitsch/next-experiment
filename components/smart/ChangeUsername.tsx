import FormControl from "@material-ui/core/FormControl";
import FormHelperText from "@material-ui/core/FormHelperText";
import gql from "graphql-tag";
import React from "react";
import { Mutation, MutationFn } from "react-apollo";

import {
  ChangeUsername,
  ChangeUsernameVariables,
} from "../../query-types/ChangeUsername";
import TextFieldWithEnterToSubmit from "../dumb/TextFieldWithEnterToSubmit";

const CHANGE_USERNAME = gql`
  mutation ChangeUsername($username: String!) {
    changeUsername(username: $username) {
      username
      id
    }
  }
`;

class ChangeUsernameComponent extends React.Component {
  public render() {
    return (
      <Mutation mutation={CHANGE_USERNAME}>
        {(
          changeUsername: MutationFn<ChangeUsername, ChangeUsernameVariables>,
          { error },
        ) => {
          let errorInfo = [];
          if (error) {
            errorInfo = error.graphQLErrors
              .map(e => e.extensions.exception.details)
              .reduce((acc, val) => acc.concat(val), [])
              .map(e => <FormHelperText key={e}>{e}</FormHelperText>);
          }
          return (
            <FormControl error={!!error}>
              <TextFieldWithEnterToSubmit
                id="change-username-field"
                doClear
                onSubmit={username => {
                  changeUsername({ variables: { username } });
                }}
                textFieldProps={{
                  error: !!error,
                  label: "Enter new Username",
                }}
              />
              {errorInfo}
            </FormControl>
          );
        }}
      </Mutation>
    );
  }
}

export default ChangeUsernameComponent;
