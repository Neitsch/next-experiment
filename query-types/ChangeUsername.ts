/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: ChangeUsername
// ====================================================

export interface ChangeUsername_changeUsername {
  username: string | null;
  id: string;
}

export interface ChangeUsername {
  changeUsername: ChangeUsername_changeUsername;
}

export interface ChangeUsernameVariables {
  username: string;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
