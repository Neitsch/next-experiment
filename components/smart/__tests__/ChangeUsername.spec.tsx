import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import ChangeUsername from "../ChangeUsername";

jest.unmock("../ChangeUsername");
jest.unmock("../../dumb/TextFieldWithEnterToSubmit");

describe("ChangeUsername", () => {
  describe("renders", () => {
    it("base", () => {
      const rendered = shallow(<ChangeUsername />);
      expect(rendered).toMatchSnapshot();
    });
    it("mutation", () => {
      const changeUsername = jest.fn();
      const baseRendered = shallow(<ChangeUsername />);
      const mutationRendered = baseRendered
        .props()
        .children(changeUsername, {});
      expect(mutationRendered).toMatchSnapshot();
    });
    it("submit setup", () => {
      const changeUsername = jest.fn();
      const baseRendered = shallow(<ChangeUsername />);
      const mutationRendered = shallow(
        baseRendered.props().children(changeUsername, {}),
      ).find("#change-username-field");
      mutationRendered.props().onSubmit("Test");
      expect(changeUsername).toHaveBeenLastCalledWith({
        variables: { username: "Test" },
      });
    });
    it("Error display", () => {
      const changeUsername = jest.fn();
      const baseRendered = shallow(<ChangeUsername />);
      const mutationRendered = shallow(
        baseRendered.props().children(changeUsername, {
          error: {
            graphQLErrors: [
              { extensions: { exception: { details: ["Test Details"] } } },
            ],
          },
        }),
      );
      expect(mutationRendered).toMatchSnapshot();
    });
  });
});
