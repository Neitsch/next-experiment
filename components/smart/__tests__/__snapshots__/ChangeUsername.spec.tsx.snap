// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeUsername renders Error display 1`] = `
<FormControl
  classes={
    Object {
      "fullWidth": "MuiFormControl-fullWidth-4",
      "marginDense": "MuiFormControl-marginDense-3",
      "marginNormal": "MuiFormControl-marginNormal-2",
      "root": "MuiFormControl-root-1",
    }
  }
  component="div"
  disabled={false}
  error={true}
  fullWidth={false}
  margin="none"
  required={false}
>
  <TextFieldWithEnterToSubmit
    doClear={true}
    id="change-username-field"
    onSubmit={[Function]}
    textFieldProps={
      Object {
        "error": true,
        "label": "Enter new Username",
      }
    }
  />
  <WithStyles(FormHelperText)
    key="Test Details"
  >
    Test Details
  </WithStyles(FormHelperText)>
</FormControl>
`;

exports[`ChangeUsername renders base 1`] = `
<mockConstructor>
  <Component />
</mockConstructor>
`;

exports[`ChangeUsername renders mutation 1`] = `
<WithStyles(FormControl)
  error={false}
>
  <TextFieldWithEnterToSubmit
    doClear={true}
    id="change-username-field"
    onSubmit={[Function]}
    textFieldProps={
      Object {
        "error": false,
        "label": "Enter new Username",
      }
    }
  />
</WithStyles(FormControl)>
`;
